contrib/babelfishpg_tsql/src/pl_exec-2.c

```diff
@@ -2493,6 +2513,159 @@ is_char_identpart(char c)
 			(c >= '0' && c <= '9'));
 }
 
+/*
+ * Put delimiters around a T-SQL variable/parameter that is
+ * named '@@var' or contains a hash, e.g. '@var#'.
+ * Without delimiters, the backend will raise an error.
+ * This is used for the parameter argument of sp_executesql, so the input
+ * string may contain multiple names, e.g.: @par1 int, @par2 varchar(20), ...
+ * This function calls palloc() to allocate a new string and returns a pointer
+ * to this string.
+ */
+static char *
+delimit_tsql_atatuservar(const char *src)
+{
+	char *s = (char *) src;
+	char *varname_start = NULL;
+	bool add_delimiter = false;
+	
+	/* 
+	 * Reserving twice the amount of space of the input string: since the shortest possible
+	 * parameter definition is 5 characters ('@@p x' , where x would be the type), this will 
+	 * always be enough for adding delimiters.
+	 * Note that there can be multiple parameter names in the input string.
+	 */
+	char *result = (char *) palloc(sizeof(char)*strlen(src)*2);	
+	char *tgt = result;
+
+	while (*s)
+	{
+		/* Look for start of variable name, which is always '@' */
+		if (*s != '@')
+		{
+			*tgt++ = *s++;
+			continue;
+		}
+
+		/* Start of variable name found */
+		add_delimiter = false;
+
+		varname_start = s;
+
+		/* Name starting with @@ */
+		if (*(s+1))
+		{
+			if (*(s+1) == '@')
+			{
+				add_delimiter = true;
+			}
+		}
+
+		/* Find end of variable name */
+		while (*s)
+		{
+			/* Allowed chars in @variable name: */
+			if (isalpha(*s) || isdigit(*s) || (*s == '_') || (*s == '@') || (*s == '#'))
```

Comment:
what might happen here for names in different languages/collations, ie non latin chars?

Reply:
we should better use [is_identifier_char()](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/f29e512f0f8ec1e30cc310cb95980ed6025be151/contrib/babelfishpg_tsql/src/hooks.c#L1949) if that helps here

Reply:
`is_identifier_char()` does not check for `@` characters (though it probably should since these are valid in T-SQL). The test above has been relocated into a new function `is_variable_name_char()` . Both these functions do actually not test for letters in non-ASCII ranges (e.g. Chinese etc.). It would seem to make sense to unify these two functions and to add support for beyond-ASCII characters as well.  However this exceeds the scope of the current fix and should be dealt with separately.

For example, the following is valid in T-SQL but currently fails in Babefish because the Chinese characters are not recognized by `is_variable_name_char()` .
```
1> -- variable name with two '@@' 
2> declare @@地点 int = 124 
3> select @@地点 
4> go 
Msg 33557097, Level 16, State 1, Server BABELFISH, Line 3 
function sys.地点() does not exist 
```

It should be noted that regular identifiers (e.g. table names) and variables like the above example but with only one `@`, are currently handled correctly by Babelfish because these names are just handed off to the PG backend where the backend lexer handles these, and the lexer is Unicode-aware. For cases where Babelfish needs to explicitly check the characters in the name though `is_identifier_char()`  or `is_variable_name_char()` , this will currently result in an error for such names.

Reply:
thanks for looking into this Rob. Agree that this exceeds the scope of this change.